### PR TITLE
RUN-1392: fix events recorder crashes 2

### PR DESCRIPTION
**Pull request by Domiii**
Fixing events yet again!
* Issue: https://linear.app/replay/issue/RUN-1392/fix-events-recorder-crashes
* Last analysis: https://linear.app/replay/issue/RUN-1392#comment-b25ddaff
* Test Results (150):
  * [This is the version of the file](https://github.com/replayio/chromium/blob/a852f2fd6dc632c9785e1020218b26983ef4ff88/third_party/blink/renderer/bindings/core/v8/record_replay_events.cc#L61) that I tested it on with events enabled unconditionally.
  * [These are the result logs](https://github.com/replayio/build-test-dashboard/blob/master/8c1c572a-04c0-4046-b5a1-a6b383e34821): `131 TestPassed, 15 TestFailed`

### CODE REVIEW

File: third_party/blink/renderer/bindings/core/v8/record_replay_events.cc [hunk 1]
```diff
@@ -108,7 +108,9 @@ UserEventProbe::UserEventProbe()
 {}
 
 UserEventProbe::~UserEventProbe() {
-  ReplayNotifyAfterEvent(name_, event_target_, is_callback_);
+  if (!name_.empty()) {  // only trigger this if the probe was initialized
+    ReplayNotifyAfterEvent(name_, event_target_, is_callback_);
+  }
```
Comment (bhackett1024) [inline third_party/blink/renderer/bindings/core/v8/record_replay_events.cc:113]: Hmm, this code is written in a way that's hard to follow.  Calls to ReplayNotifyBeforeEvent and ReplayNotifyAfterEvent should be symmetric.  With this fix we call BeforeEvent unconditionally and AfterEvent conditionally, because BeforeEvent checks for a missing name but AfterEvent doesn't.

There are several ways to improve this but I think the different tests in BeforeEvent vs. AfterEvent are the main problem --- they led to this bug and it's hard to see at a glance whether they are equivalent.  I think it would be better if we had a ShouldNotifyEvent function which is called by both functions.  That would allow removing the gIsEventInFlight variable as then it is only used in an assertion.

**Review by bhackett1024 — CHANGES_REQUESTED** [review]

File: third_party/blink/renderer/bindings/core/v8/record_replay_events.cc (hunk 1 above)
Comment (Domiii) [reply third_party/blink/renderer/bindings/core/v8/record_replay_events.cc]: Fixed.
* Then [ran another test](https://github.com/replayio/build-test-dashboard/blob/master/272aafbd-64b9-4ac6-8e4d-f8eae21bc514) with 26 x `Passed`, 2 x `Failed`.
* Then added back the `event flag` check.

File: third_party/blink/renderer/bindings/core/v8/record_replay_events.cc [hunk 2]
```diff
@@ -53,14 +53,19 @@ static void ReplayNotifyAfterEvent(const String& eventName,
                                    EventTarget* eventTarget = nullptr,
                                    bool isCallback = false);
 
+bool ShouldNotifyEvent(const String& eventName) {
+  return IsRecordingOrReplaying() && !AreEventsDisallowed() &&
+         !eventName.empty() &&
+          // Disabled by default (RUN-1251)
+          !FeatureEnabled("disable-collect-events") &&
```
Comment (bhackett1024) [inline third_party/blink/renderer/bindings/core/v8/record_replay_events.cc:60]: indentation

**Review by bhackett1024 — APPROVED** [review]
Thanks!
